plugins/pixart-rf/fu-pxi-wireless-device.c

```diff
@@ -0,0 +1,606 @@
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_wireless_device_write_payload (FuDevice *device, FuChunk *chk, GError **error)
+{
+	FuPxiWirelessDevice *self = FU_PXI_WIRELESS_DEVICE (device);
+	FuPxiReceiverDevice *parent = fu_pxi_wireless_device_get_parent (device ,error);
+	guint8 buf[FU_PXI_RECEIVER_DEVICE_OTA_BUF_SZ] = { 0x0 };
+	guint8 status = 0x0;
+	g_autoptr(GByteArray) ota_cmd = g_byte_array_new ();
+	g_autoptr(GByteArray) receiver_cmd = g_byte_array_new ();
+	g_autoptr(GTimer) timer = g_timer_new ();
+
+	/* ota write payload command */
+	fu_byte_array_append_uint8 (ota_cmd, fu_chunk_get_data_sz (chk));			/* ota command length */
+	g_byte_array_append (ota_cmd, fu_chunk_get_data (chk), fu_chunk_get_data_sz (chk));	/* payload content */
+
+	/* increase the serial number */
+	self->sn++;
+
+	/* get pixart wireless module for ota write payload command */
+	if (!fu_pxi_composite_receiver_cmd (FU_PXI_DEVICE_CMD_FW_OTA_PAYLOAD_CONTENT,
+							self->sn,
+							self->model.target,
+							receiver_cmd,
+							ota_cmd, error))
+		return FALSE;
+
+	if (!fu_pxi_wireless_device_set_feature (FU_DEVICE (parent),
+						 receiver_cmd->data,
+						 receiver_cmd->len,
+						 error))
+		return FALSE;
+
+	/* delay for wireless module device get command response*/
+	g_usleep (50 * 1000);
+
+	if (!fu_pxi_wireless_device_get_cmd_response (self, buf, sizeof(buf), error))
+			return FALSE;
+
+	if (!fu_common_read_uint8_safe (buf, sizeof(buf), 0x5, &status, error))
+		return FALSE;
+
+	if (status != OTA_RSP_OK) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "cmd rsp check fail: %s [0x%02x]",
+			     fu_pxi_receiver_cmd_result_to_string (status),
+			     status);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_wireless_device_write_chunk (FuDevice *device, FuChunk *chk, GError **error)
+{
+	FuPxiWirelessDevice *self = FU_PXI_WIRELESS_DEVICE (device);
+	guint16 checksum;
+	guint32 prn = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* send create fw object command */
+	if (!fu_pxi_wireless_device_fw_object_create (device, chk, error))
+		return FALSE;
+
+	/* write payload */
+	chunks = fu_chunk_array_new (fu_chunk_get_data (chk),
+				     fu_chunk_get_data_sz (chk),
+				     fu_chunk_get_address (chk),
+				     0x0, self->fwstate.mtu_size);
+
+	/* calculate checksum of chunk */
+	checksum = fu_pxi_common_sum16 (fu_chunk_get_data (chk), fu_chunk_get_data_sz (chk));
+	self->fwstate.checksum += checksum;
+
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk2 = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_wireless_device_write_payload (device, chk2, error))
+			return FALSE;
+		prn++;
+		/* check crc at fw when PRN over threshold write or
+		 * offset reach max object sz or write offset reach fw length */
+		if (prn >= self->fwstate.prn_threshold || i == chunks->len - 1) {
```

Comment:
I always get worried about compiler order of operations, I would feel better with it being `i == (chunks->len -1)`

Reply:
fixed it.